### PR TITLE
Add test for causal only variant of some encoder-decoder models

### LINKED ISSUE #48745: `RoFormerForCausalLM`, `BigBirdForCausalLM`, `RemBertForCausalLM`, `MegatronBertForCausalLM` attend to future tokens since v5.9.0 — CLM training loss collapses
### System Info

```
- `transformers` version: 5.18.0.dev0
- Platform: Linux-7.0.0-31-generic-x86_64-with-glibc2.39
- Python version: 3.10.20
- Huggingface_hub version: 1.31.0
- Safetensors version: 0.8.0
- Accelerate version: 1.13.0
- Accelerate config: not found
- DeepSpeed version: not installed
- PyTorch version (accelerator?): 2.12.1+cu130 (CUDA)
- Using distributed or parallel set-up in script?: no
- Using GPU in script?: no, the script below is CPU-only
- GPU type: NVIDIA GeForce RTX 3060
```

### Who can help?

@vasqu @Cyrilvallez

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [x] My own task or dataset (give details below)

### Reproduction

Training `RoFormerForCausalLM` (`is_decoder=True`) as a plain causal LM, the loss fell to almost zero within a few hundred steps on random tokens. The decoder is not causal: each position sees the tokens after it, so its label is visible. `BertLMHeadModel` is fine; these four are not.

`main` at `5474a55e92`, CPU:

```python
import torch
from transformers import (
    BertConfig, BertLMHeadModel,
    BigBirdConfig, BigBirdForCausalLM,
    MegatronBertConfig, MegatronBertForCausalLM,
    RemBertConfig, RemBertForCausalLM,
    RoFormerConfig, RoFormerForCausalLM,
)

torch.manual_seed(0)
kw = dict(vocab_size=99, hidden_size=32, num_hidden_layers=2, num_attention_heads=4, intermediate_size=37, is_decoder=True)
models = {
    "bert": BertLMHeadModel(BertConfig(**kw)),
    "big_bird": BigBirdForCausalLM(BigBirdConfig(attention_type="original_full", **kw)),
    "megatron_bert": MegatronBertForCausalLM(MegatronBertConfig(**kw)),
    "rembert": RemBertForCausalLM(RemBertConfig(input_embedding_size=32, output_embedding_size=32, **kw)),
    "roformer": RoFormerForCausalLM(RoFormerConfig(embedding_size=32, **kw)),
}

# 1. Do the logits at positions 0..6 change when only token 7 changes?
ids = torch.randint(3, 99, (1, 8))
other = ids.clone()
other[0, -1] = (ids[0, -1] + 1) % 96 + 3
for name, model in models.items():
    model.eval()
    with torch.no_grad():
        delta = (model(input_ids=ids).logits[0, :-1] - model(input_ids=other).logits[0, :-1]).abs().max().item()
    print(f"{name:14} earlier logits move by {delta:.4f} when the last token changes")

# 2. Train as a causal LM on fresh random tokens: nothing is predictable, so the loss cannot go below ln(96) = 4.56
for name in ("bert", "roformer"):
    model = models[name].train()
    opt = torch.optim.Adam(model.parameters(), lr=3e-3)
    for step in range(600):
        ids = torch.randint(3, 99, (8, 16))
        loss = model(input_ids=ids, labels=ids).loss
        opt.zero_grad(); loss.backward(); opt.step()
    print(f"{name:14} loss after 600 steps on random tokens: {loss.item():.2f}")
```

```
bert           earlier logits move by 0.0000 when the last token changes
big_bird       earlier logits move by 0.0007 when the last token changes
megatron_bert  earlier logits move by 0.0149 when the last token changes
rembert        earlier logits move by 0.0009 when the last token changes
roformer       earlier logits move by 0.0011 when the last token changes
bert           loss after 600 steps on random tokens: 4.57
roformer       loss after 600 steps on random tokens: 0.11
```

Pretrained weights (`from_pretrained(..., is_decoder=True)`): `junnyu/roformer_chinese_base` as `RoFormerForCausalLM` moves earlier logits by 6.5 when a later token changes, `google/bigbird-roberta-base` as `BigBirdForCausalLM` by 9.0. Greedy `generate()` with and without `use_cache` also disagree.

`RoFormerModel.forward` and the other three always call `create_bidirectional_mask` (`modeling_roformer.py:727`), where `BertModel` branches on `is_decoder` and calls `create_causal_mask` (`modeling_bert.py:696-708`). Before #43924 all went through `get_extended_attention_mask`, which had that branch built in; on the commit before it the script prints 0.0000 everywhere. A v5.9.0 regression, invisible to CI because these four are still excluded from the generation tests and `test_decoder_model_past_with_large_inputs` compares the embedding output, which no mask can change.

### Expected behavior

With `is_decoder=True` the four should be causal, as up to v5.8 and as BERT still is: a change to token *k* leaves the logits before *k* untouched, and cached and cacheless `generate()` agree.

Nobody seems to have reported it — a loss that goes down too well is not an error — but any `*ForCausalLM` fine-tune of these families with `is_decoder=True` since May, and any `EncoderDecoderModel` with a RoFormer, RemBert or MegatronBert decoder, trains with the labels visible.

One caveat: the `junnyu/roformer_chinese_sim_char_*` checkpoints are SimBERT/UniLM-trained (bidirectional prompt), and the bug happens to give cached `generate()` exactly that, so they generate better on ≥ 5.9 than on 5.8 or with the fix — a prefix-LM mask, not `is_decoder` semantics.

***I have the fix (BERT's `is_decoder` branch in each of the four) and a regression test ready locally — red on `main` per model, green with the fix. Happy to open the PR if a maintainer wants it.***

### PR DESCRIPTION
**Pull request by nandan2003**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48760) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48760)

# What does this PR do?
Fixes #48745

When configured as decoders (`is_decoder=True`), `RoFormer`, `BigBird`, `MegatronBert`, and `RemBert` unconditionally constructed bidirectional attention masks via `create_bidirectional_mask(...)`. This allowed future tokens to leak into earlier token representations during autoregressive causal language modeling.

This PR:
 
   - Imports `create_causal_mask` from `...masking_utils` across all four model implementations.
 
   - Branches on `if self.config.is_decoder` to build a causal mask when acting as a decoder (matching the reference implementation in `BertModel`), falling back to bidirectional masking otherwise.
 
   - Adds regression assertions in the test suites across all four models verifying that modifying future tokens produces zero change in earlier token hidden states when `is_decoder=True`.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://huggingface.co/docs/transformers/contributing) and the
      [Pull Request](https://huggingface.co/docs/transformers/pr_checks) checks?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes according to the [guidelines](https://github.com/huggingface/transformers/tree/main/docs)?
- [x] Did you write any new necessary [tests](https://huggingface.co/docs/transformers/testing)?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

<!-- Your PR will be replied to more quickly if you can figure out the right person to tag with @

 If you know how to use git blame, that is the easiest way, otherwise, here is a rough guide of **who to tag**.
 Please tag fewer than 3 people.

@Cyrilvallez @vasqu

### CODE REVIEW

**Comment by nandan2003** [comment]
@vasqu @Cyrilvallez

**Review by Rovyxel-Dev — APPROVED** [review]

**Review by vasqu — COMMENTED** [review]
Sorry we solve the issue in #48785 but it's the correct fix. Appreciate it!

**Comment by nandan2003** [comment]
Thanks @vasqu. Glad to hear the fix was right

Also, since this was opened earlier, already has passing CI, and includes the full regression test suite across all four models (which #48785 is currently missing), would it be simpler to merge this one directly?

Either way works for me. If you prefer landing #48785, I'm also very happy to strip the model changes here and keep this PR just for the regression tests!

**Comment by vasqu** [comment]
We will probably merge the other PR first, then we can check the tests here

**Comment by nandan2003** [comment]
Sounds great! Ping me as soon as it merges and I'll rebase right away.

**Comment by Cyrilvallez** [comment]
I actually opened the other cause this one was a bit verbose and I wasn't sure about the tests, but yeah we can check afterwards and see to maybe keep the tests

**Comment by nandan2003** [comment]
Synced with main and kept the regression tests

File: tests/models/big_bird/test_modeling_big_bird.py [hunk 1]
```diff
@@ -233,6 +233,17 @@ def create_and_check_model_as_decoder(
         result = model(input_ids, attention_mask=input_mask, token_type_ids=token_type_ids)
         self.parent.assertEqual(result.last_hidden_state.shape, (self.batch_size, self.seq_length, self.hidden_size))
 
+        # Check that decoder does not attend to future tokens
+        if input_ids.shape[1] > 1:
```
Comment (vasqu) [inline tests/models/big_bird/test_modeling_big_bird.py]: imo a separate would be nicer and do we really need the input ids shape check -> maybe would be an indicator to skip the new test then

**Review by vasqu — APPROVED** [review]
Yea I think this is fine now, maybe a reference to the issue why we add the test

@Cyrilvallez if you want to take a look as well otherwise ill merge tomorrow or so 🤗

File: tests/models/big_bird/test_modeling_big_bird.py [hunk 2]
```diff
@@ -524,6 +549,10 @@ def test_model_as_decoder_with_default_input_mask(self):
             encoder_attention_mask,
         )
 
+    def test_decoder_causal_mask(self):
+        config_and_inputs = self.model_tester.prepare_config_and_inputs_for_decoder()
+        self.model_tester.create_and_check_decoder_causal_mask(*config_and_inputs)
```
Comment (Cyrilvallez) [inline tests/models/big_bird/test_modeling_big_bird.py]: Nit but let's not have 2 different functions, let's have everything here as a self-contained test!

**Review by Cyrilvallez — APPROVED** [review]
Alright, would just like to collapse everything as a single test for each model, instead of having a test calling another method

**Comment by nandan2003** [comment]
Yup!

**Comment by nandan2003** [comment]
Thankyou @Cyrilvallez and @vasqu
